### PR TITLE
implement details on get method for user quota

### PR DESCRIPTION
**Pull request by mbaechler**

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/SerializableQuota.java [hunk 1]
```diff
@@ -34,9 +34,14 @@
     public static final long UNLIMITED = -1;
 
     public static <U extends QuotaValue<U>> SerializableQuota<U> newInstance(Quota<U> quota) {
-        return new SerializableQuota<>(new SerializableQuotaValue<>(quota.getMax()), getUsed(quota.getUsed(), SerializableQuotaValue::new));
+        return newInstance(quota.getMax(), quota.getUsed());
     }
 
+    public static <U extends QuotaValue<U>> SerializableQuota<U> newInstance(U max, U used) {
```
[the remark this answers is not in the document]
Comment (rouazana) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/SerializableQuota.java]: I guess you don't want to add strong typing of thse? :)

[the remark this answers is not in the document]
Comment (rouazana) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/SerializableQuota.java]: can you use this factory in GetMailboxesMethod test please?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/Quota.java [hunk 2]
```diff
@@ -25,29 +25,29 @@
 
 public class Quota<T extends QuotaValue<T>> {
 
-    public static <T extends QuotaValue<T>> Quota<T> quota(T used, T max) {
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/Quota.java]: can you explain the rationals? Why do you think limit is a better naming than max?

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java [hunk 3]
```diff
@@ -415,13 +444,14 @@ public void getQuotaShouldReturnSizeWhenNoCount() throws Exception {
             given()
                 .get(QUOTA_USERS + "/" + BOB.asString())
             .then()
+                .log().all()
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java]: debug detected

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/Quota.java (hunk 2 above)
Comment (mbaechler) [reply mailbox/api/src/main/java/org/apache/james/mailbox/model/Quota.java]: Of course : when there was a single max value, max is ok.
But then I introduce the "maxByScope" and max(a, b) returns the biggest value whereas limit(a, b) has no predefined math expectation, so I decided to use limitByScope.
For consistency, I changed max field to limit.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/SerializableQuota.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbaechler) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/SerializableQuota.java]: > it seems used and max are not in the order we use for other methods

right

> I guess you don't want to add strong typing of thse? :)

like introducing a `QuotaLimit` type ? No

> can you use this factory in GetMailboxesMethod test please?

It's not used there

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java [hunk 4]
```diff
@@ -102,4 +107,22 @@ public void removeDefaultMaxMessage() {
...
             .orElseGet(Throwing.supplier(this::getDefaultMaxMessage).sneakyThrow());
     }
+
+    @Override
+    public Map<Quota.Scope, QuotaCount> listMaxMessagesDetails(QuotaRoot quotaRoot) {
+        return Stream.of(
+                Pair.of(Quota.Scope.User, dao.getMaxMessage(quotaRoot)),
+                Pair.of(Quota.Scope.Global, dao.getDefaultMaxMessage()))
+            .filter(pair -> pair.getValue().isPresent())
+            .collect(Collectors.toMap(Pair::getKey, value -> value.getValue().get()));
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java]: Why not `Guavate.toImmutableMap(...)` ?

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java [hunk 5]
```diff
@@ -102,4 +107,22 @@ public void removeDefaultMaxMessage() {
...
+    }
+
+    @Override
+    public Map<Quota.Scope, QuotaSize> listMaxStorageDetails(QuotaRoot quotaRoot) {
+        return Stream.of(
+                Pair.of(Quota.Scope.User, dao.getMaxStorage(quotaRoot)),
+                Pair.of(Quota.Scope.Global, dao.getDefaultMaxStorage()))
+            .filter(pair -> pair.getValue().isPresent())
+            .collect(Collectors.toMap(Pair::getKey, value -> value.getValue().get()));
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java]: Idem

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/quota/JPAPerUserMaxQuotaManager.java [hunk 6]
```diff
@@ -91,6 +96,25 @@ public void setDefaultMaxMessage(QuotaCount defaultMaxMessageCount) {
                 .orElseGet(this::getDefaultMaxMessage);
     }
 
+    @Override
+    public Map<Quota.Scope, QuotaCount> listMaxMessagesDetails(QuotaRoot quotaRoot) {
+        return Stream.of(
+            Pair.of(Quota.Scope.User, dao.getMaxMessage(quotaRoot)),
+            Pair.of(Quota.Scope.Global, dao.getDefaultMaxMessage()))
+        .filter(pair -> pair.getValue().isPresent())
+        .collect(Collectors.toMap(Pair::getKey, value -> value.getValue().get()));
```
Comment (chibenwa) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/quota/JPAPerUserMaxQuotaManager.java]: Idem

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/quota/JPAPerUserMaxQuotaManager.java [hunk 7]
```diff
@@ -91,6 +96,25 @@ public void setDefaultMaxMessage(QuotaCount defaultMaxMessageCount) {
...
+
```
Comment (chibenwa) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/quota/JPAPerUserMaxQuotaManager.java]: Extra line

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/quota/JPAPerUserMaxQuotaManager.java [hunk 8]
```diff
@@ -91,6 +96,25 @@ public void setDefaultMaxMessage(QuotaCount defaultMaxMessageCount) {
...
+    }
+
+    @Override
+    public Map<Quota.Scope, QuotaSize> listMaxStorageDetails(QuotaRoot quotaRoot) {
+        return Stream.of(
+            Pair.of(Quota.Scope.User, dao.getMaxStorage(quotaRoot)),
+            Pair.of(Quota.Scope.Global, dao.getDefaultMaxStorage()))
+        .filter(pair -> pair.getValue().isPresent())
+        .collect(Collectors.toMap(Pair::getKey, value -> value.getValue().get()));
```
Comment (chibenwa) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/quota/JPAPerUserMaxQuotaManager.java]: Idem

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/quota/InMemoryPerUserMaxQuotaManager.java [hunk 9]
```diff
@@ -64,6 +68,24 @@ public void setDefaultMaxMessage(QuotaCount maxMessage) throws MailboxException
         return Optional.of(max);
     }
 
+    @Override
+    public Map<Quota.Scope, QuotaCount> listMaxMessagesDetails(QuotaRoot quotaRoot) {
+        return Stream.of(
+                Pair.of(Quota.Scope.User, Optional.ofNullable(userMaxMessage.get(quotaRoot.getValue()))),
+                Pair.of(Quota.Scope.Global, maxMessage))
+            .filter(pair -> pair.getValue().isPresent())
+            .collect(Collectors.toMap(Pair::getKey, value -> value.getValue().get()));
```
Comment (chibenwa) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/quota/InMemoryPerUserMaxQuotaManager.java]: idem

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/quota/InMemoryPerUserMaxQuotaManager.java [hunk 10]
```diff
@@ -64,6 +68,24 @@ public void setDefaultMaxMessage(QuotaCount maxMessage) throws MailboxException
...
+    }
+
+    @Override
+    public Map<Quota.Scope, QuotaSize> listMaxStorageDetails(QuotaRoot quotaRoot) {
+        return Stream.of(
+            Pair.of(Quota.Scope.User, Optional.ofNullable(userMaxStorage.get(quotaRoot.getValue()))),
+            Pair.of(Quota.Scope.Global, maxStorage))
+            .filter(pair -> pair.getValue().isPresent())
+            .collect(Collectors.toMap(Pair::getKey, value -> value.getValue().get()));
```
Comment (chibenwa) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/quota/InMemoryPerUserMaxQuotaManager.java]: Idem

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java [hunk 11]
```diff
@@ -387,8 +388,36 @@ public void getQuotaShouldReturnBothWhenValueSpecified() throws Exception {
...
+
+    @Test
+    public void getQuotaShouldReturnOnlySpecifiedValues() throws Exception {
+        maxQuotaManager.setDefaultMaxStorage(QuotaSize.size(1111));
+        maxQuotaManager.setMaxMessage(userQuotaRootResolver.forUser(BOB), QuotaCount.count(52));
+
+        JsonPath jsonPath =
+            given()
+                .get(QUOTA_USERS + "/" + BOB.asString())
+                .then()
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java]: Ident? By convention, `then()` should be on the same line than `given()`...

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java [hunk 12]
```diff
@@ -439,8 +468,8 @@ public void getQuotaShouldReturnBothWhenNoSize() throws Exception {
                 .extract()
                 .jsonPath();
 
-        assertThat(jsonPath.getObject(SIZE, Long.class)).isNull();
-        assertThat(jsonPath.getLong(COUNT)).isEqualTo(maxMessage);
+        assertThat(jsonPath.getObject("user." + SIZE, Long.class)).isNull();
+        assertThat(jsonPath.getLong("user." + COUNT)).isEqualTo(maxMessage);
     }
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java]: IMO we are missing at least one test.

How does the `GET` answer looks like when we don't have user specific quota specifications

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java (hunk 4 above)
Comment (mbaechler) [reply mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraPerUserMaxQuotaManager.java]: will do

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java (hunk 11 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java]: ok

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java (hunk 12 above)
Comment (mbaechler) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java]: ok, added

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java [hunk 13]
```diff
@@ -420,6 +420,28 @@ public void getQuotaShouldReturnOnlySpecifiedValues() throws Exception {
...
 
+    @Test
+    public void getQuotaShouldReturnGlobalValuesWhenNoUserValuesDefined() throws Exception {
+        maxQuotaManager.setDefaultMaxStorage(QuotaSize.size(1111));
+        maxQuotaManager.setDefaultMaxMessage(QuotaCount.count(12));
+
+        JsonPath jsonPath =
+            given()
+                .get(QUOTA_USERS + "/" + BOB.asString())
+                .then()
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java]: Indent

Comment (mbaechler) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java]: sorry about that, intellij doesn't like our indent and any copy/paste break it ):

**Comment by rouazana** [comment]
test this please

**Comment by aduprat** [comment]
```
Results :

Failed tests: 
  JamesSpringContextTest.springShouldLoadAndAddOnlyOneQuotaUpdaterListener Multiple Failures (2 failures)
	Error creating bean with name 'exporter' defined in class path resource [META-INF/org/apache/james/spring-server.xml]: Initialization of bean failed; nested exception is java.lang.IllegalStateException: Autowired annotation requires at least one argument: public org.apache.james.mailbox.store.quota.NoQuotaManager()
	<no message> in java.lang.NullPointerException
Tests in error: 
  JamesAppSpringMainTest.testServer:27 » BeanCreation Error creating bean with n...
```

**Comment by mbaechler** [comment]
merged

**Pull request closed without merging**